File: packages/ogdf.js/src/entry/energybased/sm.cpp
#include <ogdf/energybased/StressMinimization.h>
#include "../main.h"

EM_PORT_API(float*) SM(int node_num, int link_num, int* source, int* target, double* edgesWeight, double* nodesX, double* nodesY, int terminationCriterion, bool fixXCoords, bool fixYCoords, bool fixZCoords, bool hasInitialLayout, bool layoutComponentsSeparately, double edgeCosts, int numberOfIterations, bool useEdgeCostsAttribute){
	node* nodes;
	Graph G;
	GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics | GraphAttributes::edgeDoubleWeight);

	nodes = new node[node_num];
	for (int i = 0; i < node_num; i++){
		nodes[i] = G.newNode();
		GA.x(nodes[i]) = nodesX[i];
		GA.y(nodes[i]) = nodesY[i];
	}

	edge e;

	for (int i = 0; i < link_num; i++) {
		e = G.newEdge(nodes[source[i]], nodes[target[i]]);
		GA.bends(e);
		GA.doubleWeight(e) = edgesWeight[i];
	}
	//LayoutModule
	StressMinimization *model = new StressMinimization();

	model->convergenceCriterion(static_cast<StressMinimization::TerminationCriterion>(terminationCriterion));
	model->fixXCoordinates(fixXCoords);
	model->fixYCoordinates(fixYCoords);
	model->fixZCoordinates(fixZCoords);
	model->hasInitialLayout(hasInitialLayout);
	model->layoutComponentsSeparately(layoutComponentsSeparately);
	model->setEdgeCosts(edgeCosts);
	model->setIterations(numberOfIterations);
	model->useEdgeCostsAttribute(useEdgeCostsAttribute);

    model->call(GA);

	float* re = (float*)malloc(node_num * 2 * 4);
    for(int i = 0; i < node_num; ++i) {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}

File: packages/ogdf.js/src/layouts/energybased/sm.js
import { PARAMETER_TYPE } from '../../utils/parameters'
import createLayout from '../creater'

const NAME = 'SM'

const OUR_PARAMETER_DEFINITION = {
    useWorker: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    }
}

const ORIGIN_PARAMETER_DEFINITION = {
    terminationCriterion: {
        type: PARAMETER_TYPE.CATEGORICAL,
        range: ['None', 'PositionDifference', 'Stress'],
        default: 'None'
    },
    fixXCoords: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    },
    fixYCoords: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    },
    fixZCoords: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    },
    hasInitialLayout: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    },
    layoutComponentsSeparately: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    },
    edgeCosts: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 100
    },
    numberOfIterations: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 200
    },
    useEdgeCostsAttribute: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false
    }
}

// double* edgesWeight, double* nodesX, double* nodesY,
const ATTRIBUTES_DEFINITION = {
    node: [
        {
            name: 'nodesX',
            mapper: (node) => ('x' in node ? node.x : 0),
            type: PARAMETER_TYPE.DOUBLE
        },
        {
            name: 'nodesY',
            mapper: (node) => ('y' in node ? node.y : 0),
            type: PARAMETER_TYPE.DOUBLE
        }
    ],
    link: [
        {
            name: 'edgesWeight',
            mapper: (link) => ('weight' in link ? link.weight : 1),
            type: PARAMETER_TYPE.DOUBLE
        }
    ],
    sequence: ['edgesWeight', 'nodesX', 'nodesY']
    // sequence: ['nodesX', 'nodesY']
}

const sm = createLayout(
    NAME,
    OUR_PARAMETER_DEFINITION,
    ORIGIN_PARAMETER_DEFINITION,
    ATTRIBUTES_DEFINITION
)
export default sm
